File: aula/Trabalhos/Trabalho Teórico 4/ultima_questao/Selecao java/Selecao.java
/**
 * Algoritmo de ordenacao por selecao
 * 
 * @author Max do Val Machado
 * @version 2 01/2015
 */

class Selecao extends Geracao {
  /**
   * Construtor.
   */
  public Selecao() {
    super();
  }

  /**
   * Construtor.
   * 
   * @param int tamanho do array de numeros inteiros.
   */
  public Selecao(int tamanho) {
    super(tamanho);
  }

  /**
   * Algoritmo de ordenacao por selecao.
   */
  static int comparacoes;
  static int movimentacoes;
  public static void selecao() {
    comparacoes = 0;
    movimentacoes = 0;
    for (int i = 0; i < (n - 1); i++) {
      int menor = i;
      for (int j = (i + 1); j < n; j++) {
        if (array[menor] > array[j]) {
          menor = j;
        }
        comparacoes++;
      }
      if(menor != i) {
        swap(menor, i);
        movimentacoes += 3;
      }
    }
    MyIO.print(movimentacoes); MyIO.print(", ");
    MyIO.print(comparacoes); MyIO.print(", ");
  }

  // monta array de tamanho (tamanhoDoArray) e executa (numeroDeExecucoes) vezes o algoritmo selecao
  // mostra na tela o tempo que demorou para executar a tarefa
  public static void executarAleatorio(int numeroDeExecucoes, int tamanhoDoArray) {
    for (int i = 0; i < numeroDeExecucoes; i++) {
      MyIO.print("[" + tamanhoDoArray + "]*" + numeroDeExecucoes + ", ");
      Selecao selecao = new Selecao(tamanhoDoArray);
      selecao.aleatorio(); // preenche array com rand
      long comeco = now(); // timestamp inicio
      selecao.selecao();   // ordena array
      long fim = now();    // timestamp fim
      MyIO.println("" + (fim - comeco) / 1.0); // milissegundos
    }
  }

  public static void executarCrescente(int numeroDeExecucoes, int tamanhoDoArray) {
    for (int i = 0; i < numeroDeExecucoes; i++) {
      MyIO.print("[" + tamanhoDoArray + "]*" + numeroDeExecucoes + ", ");
      Selecao selecao = new Selecao(tamanhoDoArray);
      selecao.crescente(); // preenche array com int crescentes
      long comeco = now(); // timestamp inicio
      selecao.selecao();   // ordena array
      long fim = now();    // timestamp fim
      MyIO.println("" + (fim - comeco) / 1.0); // milissegundos
    }
  }

  public static void executarDecrescente(int numeroDeExecucoes, int tamanhoDoArray) {
    for (int i = 0; i < numeroDeExecucoes; i++) {
      MyIO.print("[" + tamanhoDoArray + "]*" + numeroDeExecucoes + ", ");
      Selecao selecao = new Selecao(tamanhoDoArray);
      selecao.decrescente(); // preenche array com int decrescentes
      long comeco = now(); // timestamp inicio
      selecao.selecao();   // ordena array
      long fim = now();    // timestamp fim
      MyIO.println("" + (fim - comeco) / 1.0); // milissegundos
    }
  }


  public static void main(String[] args) {
    // executa o script:
    MyIO.println("Aleatorio");
    executarAleatorio(33, 100);
    executarAleatorio(33, 1000);
    executarAleatorio(33, 10000);
    
    MyIO.println("Crescente");
    executarCrescente(33, 100);
    executarCrescente(33, 1000);
    executarCrescente(33, 10000);
    
    MyIO.println("Decrescente");
    executarDecrescente(33, 100);
    executarDecrescente(33, 1000);
    executarDecrescente(33, 10000);
  }
}

File: aula/Trabalhos/Trabalho Teórico 4/ultima_questao/Selecao C/selecao.c
#include <stdio.h>
#include <time.h>
#include "selecao.h"

#define MAX_SIZE 10000

int comparacoes;
int movimentacoes;
clock_t comeco, fim;

void selecao(int* array, int n) {
  comparacoes = 0;
  movimentacoes = 0;
  for (int i = 0; i < (n - 1); i++) {
    int menor = i;
    for (int j = (i + 1); j < n; j++){
        if (array[menor] > array[j]){
          menor = j;
        }
        comparacoes++;
    }
    if(menor!=i) {
      swap(&array[menor], &array[i]);
      movimentacoes += 3;
    }
  }
  printf("%d, ", movimentacoes);
  printf("%d, ", comparacoes);
}

void executarAleatorio(int numeroDeExecucoes, int tamanhoDoArray, int* array) {
  for (int i = 0; i < numeroDeExecucoes; i++) {
    printf("[%d]*%d, ", tamanhoDoArray, numeroDeExecucoes);
    aleatorio(array, tamanhoDoArray); // preenche array com rand
    comeco = clock();             // timestamp inicio
    selecao(array, tamanhoDoArray);   // ordena array
    fim = clock();                       // timestamp fim
    printf("%lf\n", (double)(fim - comeco) / (double)(CLOCKS_PER_SEC));
  }
}

void executarCrescente(int numeroDeExecucoes, int tamanhoDoArray, int* array) {
  for (int i = 0; i < numeroDeExecucoes; i++) {
    printf("[%d]*%d, ", tamanhoDoArray, numeroDeExecucoes);
    crescente(array, tamanhoDoArray); // preenche array com int crescentes
    comeco = clock();             // timestamp inicio
    selecao(array, tamanhoDoArray);   // ordena array
    fim = clock();                       // timestamp fim
    printf("%lf\n", (double)(fim - comeco) / (double)(CLOCKS_PER_SEC));
  }
}

void executarDecrescente(int numeroDeExecucoes, int tamanhoDoArray, int* array) {
  for (int i = 0; i < numeroDeExecucoes; i++) {
    printf("[%d]*%d, ", tamanhoDoArray, numeroDeExecucoes);
    decrescente(array, tamanhoDoArray); // preenche array com decrescentes
    comeco = clock();             // timestamp inicio
    selecao(array, tamanhoDoArray);   // ordena array
    fim = clock();                       // timestamp fim
    printf("%lf\n", (double)(fim - comeco) / (double)(CLOCKS_PER_SEC));  
  }
}


int main() {
  int array[MAX_SIZE];

  aleatorio(array, 100);

  // executa o script:
  printf("Aleatorio\n");
  executarAleatorio(33, 100, array);
  executarAleatorio(33, 1000, array);
  executarAleatorio(33, 10000, array);
    
  printf("Crescente\n");
  executarCrescente(33, 100, array);
  executarCrescente(33, 1000, array);
  executarCrescente(33, 10000, array);
  
  printf("Decrescente\n");
  executarDecrescente(33, 100, array);
  executarDecrescente(33, 1000, array);
  executarDecrescente(33, 10000, array);

  return 0;
}